dynamo/preprocessing/preprocessor_utils.py

```diff
@@ -1538,3 +1534,116 @@ def pca_selected_genes_wrapper(
+
+    if len(variables) < 1:
+        main_warning("No variable to regress out")
+        return
+
+    if selected_genes is None:
+        x_prev = adata.X
+    else:
+        if selected_genes not in adata.var.keys():
+            raise ValueError(str(selected_genes) + " is not a key in adata.var")
+
+        if not (adata.var[selected_genes].dtype == bool):
+            raise ValueError(str(selected_genes) + " is not a boolean")
+
+        x_prev = adata[:, adata.var.loc[:, selected_genes]].X
+
+    x_prev = x_prev.toarray().copy()
+
+    if n_cores == 1:
+        y_new = x_prev  # for the cases when variables are more than 1.
+
+        for variable in variables:
+            if variable not in adata.obs.keys():
+                main_warning("No %s in adata.obs.keys" % variable)
+                continue
+
+            # Select the variables to remove
+            x_remove = adata.obs[variable].to_numpy().reshape(-1, 1)
+
+            # Predict the effects of the variables to remove
+            y_pred = regress_out_chunk(x_remove, y_new)
+
+            # Remove the effects of the variables from the target variable
+            y_new -= y_pred
+
+        # Predict the residuals after removing the effects of the variables and save to "X_residuals_for_pca"
+        res = regress_out_chunk(x_prev, y_new)
+
+    else:
+        from multiprocessing import Pool as ThreadPool
+
+        regress_val = np.zeros((x_prev.shape[0], x_prev.shape[1]))
+        # Split the input data into chunks for parallel processing
+        chunk_size = x_prev.shape[1] // n_cores
+
+        chunks = [x_prev[:, i * chunk_size : (i + 1) * chunk_size] for i in range(n_cores - 1)]
+        chunks.append(x_prev[:, (n_cores - 1) * chunk_size :])
+
+        # Create a pool of work processes
+        pool = ThreadPool(n_cores)
+
+        for variable in variables:
+            if variable not in adata.obs.keys():
+                main_warning("No %s in adata.obs.keys" % variable)
+                continue
+
+            x_remove = [adata.obs[variable].to_numpy().reshape(-1, 1)] * n_cores
+            results = pool.starmap(regress_out_chunk, zip(x_remove, chunks))
+            regress_val += np.hstack(results)
+
+        # Remove the effects of the variables from the target variable
+        residuals = x_prev - regress_val
+
+        chunks_res = [residuals[:, i * chunk_size : (i + 1) * chunk_size] for i in range(n_cores - 1)]
+        chunks_res.append(residuals[:, (n_cores - 1) * chunk_size :])
+
+        x_prev = [x_prev] * n_cores
+
+        results = pool.starmap(regress_out_chunk, zip(x_prev, chunks_res))
+        pool.close()
+        pool.join()
+        res = np.hstack(results)
+
+    adata.obsm[obsm_store_key] = res
+
+
+def regress_out_chunk(
+    X: Union[np.ndarray, spmatrix, list], y: Union[np.ndarray, spmatrix, list]
+) -> Union[np.ndarray, spmatrix, list]:
+    """Perform linear regression to remove the effects of given variables from a target variable.
```

Comment:
this sentence doesn't make sense. "Perform a linear regression to remove the effects of cell features (percentage of mitochondria, etc) from the gene expression matrix X." 

Make sure our documentation to have more biological/single cell genomics relevance 

Actually I think X here is the predictor the values from the obs.keys and y is the gene expression? please use better names (say obs_feature, gene_expr). X is often used to represent the gene expression matrix so that is why I have the mistake above.

Reply:
Fixed as you suggested